### ISSUE TITLE
Kinesis Async client crashes around an hour

### ISSUE THREAD

**Issue by vijayprasath99** [aws/aws-sdk-ruby#2779]
### Describe the bug

We use the the Kinesis async client with region and `Aws::AssumeRoleCredentials` on initializing. We're looping `.subscribe_to_shard` along with a following `.wait` and we're seeing the messages flow though. After an iteration (about 5 mins is how long the `.wait` runs), we'd call `.subscribe_to_shard` again with the existing client. (we did this following https://aws.amazon.com/blogs/developer/announcing-amazon-kinesis-subscribetoshard-api-support-in-the-aws-sdk-for-ruby/). But this flow breaks around about 1 hour (give or take 10 mins) where we start seeing connections fail with `Http2StreamInitializeError` for `subscribe_to_shard`. It is not able to recover. Only a crash and restart of the whole application resolves this.

I was able to debug the flow and found that I'm getting a `broken pipe` when it tries to send data for `subscribe_to_shard` (around an hour later after the application has started). On debugging this further, deep inside in `OpenSSL::Buffering#write`, I can see a broken pipe when it tries to write the stream. I hacked around and found that when I re-initialize `@connection = H2::Connection.new(options)` in `Seahorse::Client::AsyncBase` just before `subscribe_to_shard`, it recovers and successfully starts transmitting data again. 

Environment:
Ruby Version: `jruby-9.2.5.0`
Gem versions:
```  
gem "aws-sdk-kinesis", "~> 1.10.0"
 gem "http-2", "~> 0.10"
```
Bug also exists in the latest version of the gem
```
gem "aws-sdk-kinesis", "~> 1.42.0"
gem "http-2", "~> 0.11"
```

### Expected Behavior

Async client should not fail to connect on calling `subscribe_to_shard`

### Current Behavior

Async client fail to connect to kinesis on calling `subscribe_to_shard` after apx. 1 hour (give or take 10 mins) of execution

### Reproduction Steps

Initialize client with region and AssumedCredentials
loop though `.subscribe_to_shard` followed by a `.wait` (subscribe_to_shard should be called once every 5 mins)
after 1 hour of execution, `subscribe_to_shard` will fail to connect 

### Gem name ('aws-sdk', 'aws-sdk-resources' or service gems like 'aws-sdk-s3') and its version

aws-sdk-kinesis: 1.42.0

### Environment details (Version of Ruby, OS environment)

RUBY: jruby-9.2.5.0 OS: Linux, MacOS

**Comment by mullermp** [comment]
Thanks for opening an issue. Can you by chance also attach a stack trace? Is there any information in the wire log (set `http_wire_trace: true` in config) that would provide insight on why the stream is closing? One other thing to check, the 1 hour mark is suspicious, are your credentials by chance expiring every 1 hour?

**Comment by vijayprasath99** [comment]
To rule out any credentials as an issue,

First, I updated the credentials to have a duration of 900s in the initializer
```
       @role_credentials = Aws::AssumeRoleCredentials.new(
            role_arn: role_arn,
            role_session_name: role_name,
            duration_seconds: 900,
          )
 ```
 The client still was able to run for nearly 60 mins (It uses background refresh every 5 mins is what I've noticed). During the crash in a resque block, I was able to use the same `@role_credentials` object created during on startup inside another sync client and make an API call to AWS Kinesis which succeeded
 ```
 Aws::Kinesis::Client.new(credentials: @role_credentials).describe_stream_consumer(
          stream_arn: stream_arn,
          consumer_name: consumer_name
        ).consumer_description.consumer_status
 ```
The `@role_credentials` object's expiration is 15 mins out in future at the time of `subscribe_to_shard` failure.

 As a final, step, I had 2 clients, sync client and async client which use the same `@role_credentials` object. Right before calling `subscribe_to_shard`, I've invoked `describe_stream_consumer` to verify if the consumer is active. When the async client fails to subscribe, the sync client is able to still talk to Kinesis (both fns are in a loop so after a failure to subscribe, it is still able to talk to Kinesis service to describe_stream_consumer). As a final test, I manually revoke session using `Revoke all active sessions` in IAM console, thats when I saw `<Aws::Kinesis::Errors::AccessDeniedException> User: arn:aws:sts::XXXXXX is not authorized to perform:kinesis:DescribeStreamConsumer on resource: arn:aws:kinesis:XXXX  with an explicit deny in an identity-based policy`.
 
So something in the async client is failing to refresh or is different from the sync client in terms of using refreshed credentials.
 
 Here's a sample log with `http_wire_trace: true`
 ```
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x00\x00\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47898854266005913857422065123575799283833479007471403058\",\"MillisBehindLatest\":0,\"Records\":[]}\xDB\xF7\xB2~"}
2022-11-07T19:27:11+00:00 DEBUG reqid=x3v6b5c9-nklk-909c-86fe-7axxe21beeee2 event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x00\x00\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47898854266005913857422065131974206952696308159033114674\",\"MillisBehindLatest\":0,\"Records\":[]}\xV2\x44\xD32"}
2022-11-07T19:27:16+00:00 DEBUG reqid=x3v6b5c9-nklk-909c-86fe-7axxe21beeee2 event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x00\x00\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47898854266005913857422065139144345988830674137812828210\",\"MillisBehindLatest\":0,\"Records\":[]}C\xA8-["}
2022-11-07T19:27:21+00:00 DEBUG reqid=x3v6b5c9-nklk-909c-86fe-7axxe21beeee2 event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x00\x00\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47898854266005913857422065145875644952444929726174199858\",\"MillisBehindLatest\":0,\"Records\":[]}\x1X\xAZOy"}
2022-11-07T19:27:26+00:00 DEBUG reqid=x3v6b5c9-nklk-909c-86fe-7axxe21beeee2 event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>8, :type=>:goaway, :flags=>[], :stream=>0, :last_stream=>23, :error=>:no_error}
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x00\x00\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47898854266005913857422065153386701069710621132221055026\",\"MillisBehindLatest\":0,\"Records\":[]}D\xXX\x55"}
2022-11-07T19:27:31+00:00 DEBUG reqid=x3v6b5c9-nklk-909c-86fe-7axxe21beeee2 event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>0, :type=>:data, :flags=>[:end_stream], :stream=>23, :payload=>""}
2022-11-07T19:27:34+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : async_resp.wait returned, re-subscribe to shard_id: shardId-000000000003
closing connection ...
2022-11-07T19:27:34+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Kinesis EventHandler - subscribe to shard with shard_id: shardId-000000000003, starting pos: {:type=>"AT_TIMESTAMP", :timestamp=>2022-11-07 19:19:13 +0000}
2022-11-07T19:27:35+00:00 DEBUG reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:44 : KinesisAsyncClient - subscribe_to_shard consumer_arn: arn:aws:kinesis:us-east-1:12312124124124:stream/STREAM_NAME/consumer/event_dispatcher:1641410072, shard_id, shardId-000000000003
#<Seahorse::Client::Http2StreamInitializeError: Seahorse::Client::Http2StreamInitializeError>
2022-11-07T19:27:35+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Waiting on async response...
2022-11-07T19:28:55+00:00 DEBUG reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:71 : Consumer clients ACTIVE
2022-11-07T19:29:18+00:00 WARN  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Http2StreamInitializeError thrown, re-subscribe to shard. Seahorse::Client::Http2StreamInitializeError:Seahorse::Client::Http2StreamInitializeError
2022-11-07T19:29:18+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Kinesis EventHandler - subscribe to shard with shard_id: shardId-000000000003, starting pos: {:type=>"AT_TIMESTAMP", :timestamp=>2022-11-07 19:19:13 +0000}
2022-11-07T19:29:19+00:00 DEBUG reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:44 : KinesisAsyncClient - subscribe_to_shard consumer_arn: arn:aws:kinesis:us-east-1:12312124124124:stream/STREAM_NAME/consumer/event_dispatcher:1641410072, shard_id, shardId-000000000003
#<Seahorse::Client::Http2StreamInitializeError: Seahorse::Client::Http2StreamInitializeError>
2022-11-07T19:30:19+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Waiting on async response...
2022-11-07T19:43:02+00:00 DEBUG reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:71 : Consumer clients ACTIVE
2022-11-07T19:44:09+00:00 WARN  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Http2StreamInitializeError thrown, re-subscribe to shard. Seahorse::Client::Http2StreamInitializeError:Seahorse::Client::Http2StreamInitializeError
2022-11-07T19:44:09+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Kinesis EventHandler - subscribe to shard with shard_id: shardId-000000000003, starting pos: {:type=>"AT_TIMESTAMP", :timestamp=>2022-11-07 19:19:13 +0000}
2022-11-07T19:44:09+00:00 DEBUG reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:44 : KinesisAsyncClient - subscribe_to_shard consumer_arn: arn:aws:kinesis:us-east-1:12312124124124:stream/STREAM_NAME/consumer/event_dispatcher:1641410072, shard_id, shardId-000000000003
#<Seahorse::Client::Http2StreamInitializeError: Seahorse::Client::Http2StreamInitializeError>
2022-11-07T19:44:44+00:00 INFO  reqid=b7646c8as-c353-123a-1234-76f8cedc6619 /event-app/events/kinesis_event_handler.rb:37 : Waiting on async response...
```

**Comment by vijayprasath99** [comment]
To rule out any credentials as an issue,

First, I updated the credentials to have a duration of 900s in the initializer
```
       @role_credentials = Aws::AssumeRoleCredentials.new(
            role_arn: role_arn,
            role_session_name: role_name,
            duration_seconds: 900,
          )
 ```
 The client still was able to run for nearly 60 mins (It uses background refresh every 5 mins is what I've noticed). During the crash in a resque block, I was able to use the same `@role_credentials` object created during on startup inside another sync client and make an API call to AWS Kinesis which succeeded
 ```
 Aws::Kinesis::Client.new(credentials: @role_credentials).describe_stream_consumer(
          stream_arn: stream_arn,
          consumer_name: consumer_name
        ).consumer_description.consumer_status
 ```
The `@role_credentials` object's expiration was 15 mins out in future at the time of `subscribe_to_shard` failure.

 As a final, step, I had 2 clients, sync client and async client which use the same `@role_credentials` object. Right before calling `subscribe_to_shard`, I've invoked `describe_stream_consumer` to verify if the consumer is active. When the async client fails to subscribe, the sync client is able to still talk to Kinesis (both fns are in a loop so after a failure to subscribe, it is still able to talk to Kinesis service to describe_stream_consumer). As a final test, I manually revoke session using `Revoke all active sessions` in IAM console, thats when I saw `<Aws::Kinesis::Errors::AccessDeniedException> User: arn:aws:sts::XXXXXX is not authorized to perform:kinesis:DescribeStreamConsumer on resource: arn:aws:kinesis:XXXX  with an explicit deny in an identity-based policy`.
 
So something in the async client is failing to refresh or is different from the sync client in terms of using refreshed credentials.
 
 Here's a sample log with `http_wire_trace: true`
 ```
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x01\x02\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47893846166205916757422065123575799283833479007471403058\",\"MillisBehindLatest\":0,\"Records\":[]}\xDB\xF7\xB2~"}
2022-11-07T19:27:11+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x01\x02\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47893846166205916757422065131974206952696308159033114674\",\"MillisBehindLatest\":0,\"Records\":[]}\xV2\x44\xD32"}
2022-11-07T19:27:16+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x01\x02\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47893846166205916757422065139144345988830674137812828210\",\"MillisBehindLatest\":0,\"Records\":[]}C\xA8-["}
2022-11-07T19:27:21+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x01\x02\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47893846166205916757422065145875644952444929726174199858\",\"MillisBehindLatest\":0,\"Records\":[]}\x1X\xAZOy"}
2022-11-07T19:27:26+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>8, :type=>:goaway, :flags=>[], :stream=>0, :last_stream=>23, :error=>:no_error}
<- frame: {:length=>242, :type=>:data, :flags=>[], :stream=>23, :payload=>"\x00\x01\x02\v:event-type\a\x00\x15SubscribeToShardEvent\r:content-type\a\x00\x1Aapplication/x-amz-json-1.1\r:message-type\a\event{\"ContinuationSequenceNumber\":\"47893846166205916757422065153386701069710621132221055026\",\"MillisBehindLatest\":0,\"Records\":[]}D\xXX\x55"}
2022-11-07T19:27:31+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX event-app/gems/aws-sdk-core-3.166.0/lib/aws-sdk-core/event_emitter.rb:28 : RECEIVED EVENT: subscribe_to_shard_event with 0 records
<- frame: {:length=>0, :type=>:data, :flags=>[:end_stream], :stream=>23, :payload=>""}
2022-11-07T19:27:34+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : async_resp.wait returned, re-subscribe to shard_id: shardId-000000000001
closing connection ...
2022-11-07T19:27:34+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Kinesis EventHandler - subscribe to shard with shard_id: shardId-000000000001, starting pos: {:type=>"AT_TIMESTAMP", :timestamp=>2022-11-07 19:19:13 +0000}
2022-11-07T19:27:35+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:44 : KinesisAsyncClient - subscribe_to_shard consumer_arn: arn:aws:kinesis:us-east-1:1234567890123:stream/STREAM_NAME/consumer/event_dispatcher:123456789, shard_id, shardId-000000000001
#<Seahorse::Client::Http2StreamInitializeError: Seahorse::Client::Http2StreamInitializeError>
2022-11-07T19:27:35+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Waiting on async response...
2022-11-07T19:28:55+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:71 : Consumer clients ACTIVE
2022-11-07T19:29:18+00:00 WARN  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Http2StreamInitializeError thrown, re-subscribe to shard. Seahorse::Client::Http2StreamInitializeError:Seahorse::Client::Http2StreamInitializeError
2022-11-07T19:29:18+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Kinesis EventHandler - subscribe to shard with shard_id: shardId-000000000001, starting pos: {:type=>"AT_TIMESTAMP", :timestamp=>2022-11-07 19:19:13 +0000}
2022-11-07T19:29:19+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:44 : KinesisAsyncClient - subscribe_to_shard consumer_arn: arn:aws:kinesis:us-east-1:1234567890123:stream/STREAM_NAME/consumer/event_dispatcher:123456789, shard_id, shardId-000000000001
#<Seahorse::Client::Http2StreamInitializeError: Seahorse::Client::Http2StreamInitializeError>
2022-11-07T19:30:19+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Waiting on async response...
2022-11-07T19:43:02+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:71 : Consumer clients ACTIVE
2022-11-07T19:44:09+00:00 WARN  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Http2StreamInitializeError thrown, re-subscribe to shard. Seahorse::Client::Http2StreamInitializeError:Seahorse::Client::Http2StreamInitializeError
2022-11-07T19:44:09+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Kinesis EventHandler - subscribe to shard with shard_id: shardId-000000000001, starting pos: {:type=>"AT_TIMESTAMP", :timestamp=>2022-11-07 19:19:13 +0000}
2022-11-07T19:44:09+00:00 DEBUG reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:44 : KinesisAsyncClient - subscribe_to_shard consumer_arn: arn:aws:kinesis:us-east-1:1234567890123:stream/STREAM_NAME/consumer/event_dispatcher:123456789, shard_id, shardId-000000000001
#<Seahorse::Client::Http2StreamInitializeError: Seahorse::Client::Http2StreamInitializeError>
2022-11-07T19:44:44+00:00 INFO  reqid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX /event-app/events/kinesis_event_handler.rb:37 : Waiting on async response...
```

**Comment by mullermp** [comment]
`<- frame: {:length=>0, :type=>:data, :flags=>[:end_stream], :stream=>23, :payload=>""}`

This looks like the service is ending the stream for some reason. We received a support request (was this you?) that was transferred to us. We'll need to find out why we got this event.

**Comment by mullermp** [comment]
A stack trace would be very helpful, especially at the part where Http2StreamInitializeError is raised, to understand why `@h2_client.new_stream` is failing (https://github.com/igrigorik/http-2/blob/8ec79e8982cb65fb9740ed8ca6e1950774a5a533/lib/http/2/connection.rb#L112)

**Comment by alextwoods** [comment]
I'm not quite sure what your code is doing, but its possible you need to handle child shards when the current shard is ended and SubscribeToShardEvent returns a list of child shards to process next? 
 You can use https://docs.aws.amazon.com/streams/latest/dev/building-enhanced-consumers-api.html as reference for details on child shard.

**Comment by alextwoods** [comment]
I've posted a potential fix for this issue: https://github.com/aws/aws-sdk-ruby/pull/2781

Basically - it when the underlying connection's socket was closing (due to a read timeout or any other reason),  the `Connection#close!` method was calling Thread.kill(@socket_thread) before updating the `@status`. The `close!` method was being called FROM the `@socket_thread` and so killing it, prevented the status from being updated correctly, leaving the connection in an incorrect state.

Note - that you may still see errors when the underlying `@socket` closes - but that error should now be `Http2ConnectionClosedError` and you should be able to rescue that and call `async_client.new_connection` to resolve the issue.  If this is happening too frequently, you could try to increase the `connection_read_timeout` (its currently set at 60) on the async client.

**Comment by alextwoods** [comment]
This fix was released today (11-9) in [aws-sdk-core 3.167.0](https://rubygems.org/gems/aws-sdk-core/versions/3.167.0).  If this doesn't end up addressing your issue, feel free to reopen!
